Introduce the StageMode attribute to Model

Phipson opened the pull request:
#### 9ad0659e5627046e8b8464ee1fc2310c49602d03
<pre>
Introduce the StageMode attribute to Model
<a href="https://bugs.webkit.org/show_bug.cgi?id=285899">https://bugs.webkit.org/show_bug.cgi?id=285899</a>
<a href="https://rdar.apple.com/142566328">rdar://142566328</a>

Reviewed by Ada Chan.

This is the first of three PRs used to introduce the stagemode attribute to model views, which will
be used to define and control the interaction of a model. The stagemode attribute takes a string and
determines how the model behaves based on user&apos;s gesture input on the view containing the model. In
particular, we currently support 2 types of stagemode behaviors: orbit, which will allow users to rotate
the model, and none, which will disable model interaction. Subsequent PRs will be made to use this
attribute to define model interaction.

* LayoutTests/model-element/model-element-stagemode-expected.txt: Added.
* LayoutTests/model-element/model-element-stagemode.html: Added.
* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::HTMLModelElement::createModelPlayer):
Updates the stagemode attribute when the source for the model element updates
(WebCore::HTMLModelElement::attributeChanged):
Updates the stagemode attribute when the attribute changes
(WebCore::HTMLModelElement::stageMode const):
Returns the enum corresponding to the stagemode attribute as defined by the user-defined string
(WebCore::HTMLModelElement::updateStageMode):
Updates the internal variable the model player uses to track the stagemode attribute
* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/Modules/model-element/HTMLModelElement.idl:
* Source/WebCore/Modules/model-element/ModelPlayer.cpp:
(WebCore::ModelPlayer::setStageMode):
Virtual function to update the player based on the stagemode attribute
* Source/WebCore/Modules/model-element/ModelPlayer.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/HTMLAttributeNames.in:
* Source/WebCore/page/StageModeOperations.h: Added.
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.messages.in:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
(WebKit::ModelProcessModelPlayerProxy::setStageMode):
Updates the internal variable correpsonding to the stagemode attribute
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp:
(WebKit::ModelProcessModelPlayer::setStageMode):
Updates the internal variable correpsonding to the stagemode attribute
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.h:

Canonical link: <a href="https://commits.webkit.org/289830@main">https://commits.webkit.org/289830@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c3dbf9fdfbd59e851a5a2685244b6d7120302d2f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87914 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/7430 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/42303 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92811 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/38664 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/89965 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/7811 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/15606 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67872 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25610 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90916 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5983 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79545 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48240 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/87411 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5763 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/33952 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37771 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/76162 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34833 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/94679 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/15082 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/11099 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76721 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/15337 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75401 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75959 "Found 1 new API test failure: /WebKitGTK/TestWebKitPolicyClient:/webkit/WebKitPolicyClient/new-window-policy (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18716 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20347 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18770 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/8086 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/15100 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/20401 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14842 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/18287 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/16624 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on LayoutTests/model-element/model-element-stagemode.html:
```diff
@@ -0,0 +1,42 @@
+    model.stageMode = "none";
+    assert_equals(model.stageMode, "none");
+
+}, `Checking <model> stage mode value after updating`);
```
Can you add some tests where it's set to invalid values or mixed case values? I think the logic is sound for it to be defined eventually as an enumerated attribute in the HTML standard, but would be nice to confirm.

Also, ideally we make these web-platform-tests and upstream them. We'll need web-platform-tests for standardization so might as well start with those now. They'll need `.tentative` or be in a `tentative` folder, but otherwise they can be like this.

Phipson replied:
Will add additional tests. I think for the first iteration of stage mode, the proposal was that we can pass it a string as an argument. However, I agree that it makes sense that we can supply an enumerated attribute in future once we confirm the set of operations we would like to support in addition to `orbit`

annevk replied:
An enumerated attribute (a term defined in the HTML standard) uses strings so the API you have seems correct. I was talking about the standard side of things.

annevk commented on Source/WebCore/page/StageModeOperations.h:
```diff
@@ -0,0 +1,15 @@
+#if ENABLE(MODEL_PROCESS)
+namespace WebCore {
+
+enum class StageModeOperation : uint8_t {
```
bool

Phipson replied:
There is a future where we will add more operations to StageMode, which is why we are taking a string as an argument in the first place, rather than a toggle for the idl. I think that it makes sense for us to keep it as an int, unless you think that it's better for me to make it a bool and change it into an int later once we add more operations

annevk replied:
Typically we do `enum class X : bool` if it fits and then expand to uint8_t or larger as needed later on.

Phipson replied:
Got it, makes sense! Will do that then!

annevk commented on Source/WebCore/page/StageModeOperations.h:
```diff
@@ -0,0 +1,15 @@
+
+enum class StageModeOperation : uint8_t {
+    None  = 0,
+    Orbit = 1,
```
No need to assign values.

annevk commented on Source/WebCore/page/StageModeOperations.h:
```diff
@@ -0,0 +1,15 @@
+// Copyright © 2025  All rights reserved.
```
This is not the correct boilerplate.

Phipson replied:
This was the default I got when creating a new file. Is there a way to reformat the boilerplate or would I have to copy paste it myself?

annevk replied:
Copypasta is what I learned...

annevk commented on Source/WebCore/Modules/model-element/HTMLModelElement.idl:
```diff
@@ -53,6 +53,8 @@
     [Conditional=MODEL_PROCESS, EnabledBySetting=ModelProcessEnabled] Promise<undefined> pause();
     [Conditional=MODEL_PROCESS, EnabledBySetting=ModelProcessEnabled] attribute double currentTime;
+    
+    [Conditional=MODEL_PROCESS, EnabledBySetting=ModelProcessEnabled, Reflect] attribute DOMString stageMode;
```
Given that this isn't even in a draft yet, perhaps this should be behind its own setting as well?

achan00 replied:
So far it's described in the explainer: https://github.com/immersive-web/model-element/blob/main/explainer.md. We have been implementing other APIs in that explainer under the `ModelElementEnabled` and `ModelProcessEnabled` settings. 

Would we need a brand new setting just for `stagemode`?

Phipson replied:
My understanding is that stage mode will land with the model element, so I had assumed that we would use the same flag. Knowing this, do you think it still makes sense to have a dedicated `StageModeEnabled` setting?

weinig commented on Source/WebCore/Modules/model-element/HTMLModelElement.h:
```diff
@@ -42,6 +42,10 @@
 #include <wtf/UniqueRef.h>
 
+#if ENABLE(MODEL_PROCESS)
+#include <WebCore/StageModeOperations.h>
```
This should be a normal header `#include "StageModeOperations.h"` as this header is in WebCore.

Also, this enable is too broad for this. If StageMode is currently only supported by the model process, the way to go would be to add a ENABLE_MODEL_STAGE_MODE macro and set it to 1 if ENABLE_MODEL_PROCESS is set.

Phipson replied:
My understanding is that stage mode will land with the model element, so I had assumed that we would use the same flag. See https://github.com/immersive-web/model-element/blob/main/explainer.md Knowing this, do you think it still makes sense to have a dedicated `ENABLE_MODEL_STAGE_MODE` macro?

weinig commented on Source/WebKit/Scripts/webkit/messages.py:
```diff
@@ -1070,6 +1070,7 @@ def headers_for_type(type):
         'WebCore::ShouldSample': ['<WebCore/DiagnosticLoggingClient.h>'],
         'WebCore::SourceBufferAppendMode': ['<WebCore/SourceBufferPrivate.h>'],
         'WebCore::SourceBufferEvictionData': ['<WebCore/SourceBufferPrivateClient.h>'],
+        'WebCore::StageModeOperation': ['<WebCore/StageModeOperations.h>'],
```
Is this needed? I think these are only needed when the type name and header don't match.

Phipson replied:
Closing the loop, but our enum is called `StageModeOperation` whereas the file is plural and titled `StageModeOperations`. Since I think the enum itself should not be plural I think this makes sense

achan00 reviewed (approved):
Hoping to see more tests when more logic is implemented for `stage mode`.

Phipson committed:
Introduce the StageMode attribute to Model
https://bugs.webkit.org/show_bug.cgi?id=285899
rdar://142566328

Reviewed by Ada Chan.

This is the first of three PRs used to introduce the stagemode attribute to model views, which will
be used to define and control the interaction of a model. The stagemode attribute takes a string and
determines how the model behaves based on user's gesture input on the view containing the model. In
particular, we currently support 2 types of stagemode behaviors: orbit, which will allow users to rotate
the model, and none, which will disable model interaction. Subsequent PRs will be made to use this
attribute to define model interaction.

* LayoutTests/model-element/model-element-stagemode-expected.txt: Added.
* LayoutTests/model-element/model-element-stagemode.html: Added.
* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::HTMLModelElement::createModelPlayer):
Updates the stagemode attribute when the source for the model element updates
(WebCore::HTMLModelElement::attributeChanged):
Updates the stagemode attribute when the attribute changes
(WebCore::HTMLModelElement::stageMode const):
Returns the enum corresponding to the stagemode attribute as defined by the user-defined string
(WebCore::HTMLModelElement::updateStageMode):
Updates the internal variable the model player uses to track the stagemode attribute
* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/Modules/model-element/HTMLModelElement.idl:
* Source/WebCore/Modules/model-element/ModelPlayer.cpp:
(WebCore::ModelPlayer::setStageMode):
Virtual function to update the player based on the stagemode attribute
* Source/WebCore/Modules/model-element/ModelPlayer.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/HTMLAttributeNames.in:
* Source/WebCore/page/StageModeOperations.h: Added.
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.messages.in:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
(WebKit::ModelProcessModelPlayerProxy::setStageMode):
Updates the internal variable correpsonding to the stagemode attribute
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp:
(WebKit::ModelProcessModelPlayer::setStageMode):
Updates the internal variable correpsonding to the stagemode attribute
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.h:

Canonical link: https://commits.webkit.org/289830@main
